### PR TITLE
Report Aria Details In Focus Mode

### PR DESCRIPTION
**Pull request by feerrenrut**
### Link to issue number:
Follow on from nvaccess/nvda#12439

### Summary of the issue:
Reporting "has details" and summarizing the details was not supported in focus mode or content editable.

### Description of how this pull request fixes the issue:
Relies on 
- #13097
- #13096
- #13112 

Changes in this PR:
- Adds system tests for focus mode / content editable
- Exposes an in-process approach to summarize an `IAccessible` ( `getTextFromIAccessible` moved from `ia2LiveRegions.cpp` to `textFromIAccessible.h/cpp` and exposed as `nvdaInProcUtils_getTextFromIAccessible` via `IA2Support.cpp`)
- Make it easier to fetch the details relation targets from python code.
- Updates reporting of has details and details summary to work in both browse and focus mode.

### Testing strategy:
- System tests
- Alpha users

### Known issues with pull request:
Currently only a single relation can be summarized.
Support for multiple details relations will be added later, there are various UX concerns that will need to be resolved.
- [x] Braille expectation should be added to the system tests

### Change log entries:
New features
```
- Reporting of "has details" and the associated command to summarize the details relation have been updated to work in focus mode.
```

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

File: source/globalCommands.py [hunk 1]
```diff
@@ -1956,6 +1961,70 @@ def script_reportOrShowFormattingAtCaret(self, gesture):
 		elif repeats == 1:
 			self.script_showFormattingAtCaret(gesture)
 
+	@script(
+		description=_(
+			# Translators: the description for the reportDetailsSummary script.
+			"Report summary of any annotation details for this navigator position."
```
Comment (lukaszgo1) [inline source/globalCommands.py:1967]: I feel this is quite odd to report at a position of the review cursor since in both browse and focus modes you're navigating with a caret rather than with the review cursor. Has it been considered to report the details at a position of the browse mode caret if it exists and fall back to the focused object if there is no caret or user is in focus mode?

Comment (feerrenrut) [reply source/globalCommands.py:1967]: Are you referring the behavior or just this description of the behavior. The description could certainly be more precise.

If you are referring to the behavior, could you please give an example that highlights how it is odd?

Comment (lukaszgo1) [reply source/globalCommands.py]: Ive been referring to the behavior. My main problem with using position of the review cursor is that in both browse mode and content editable you're navigating with system focus or system caret, and while by default review cursor and navigator object tracks them this is optional. This means that:
1. If you're working with review cursor not following caret most of the time you need to move it from wherever you've  left it just to check annotations at your current focus / caret position and 
2. Even if you're working with the default configuration for the review cursor most of the time you may want to disable following of the navigator object for a single task, forgot about it and be caught by surprise when you want to check for annotations.

File: tests/system/robot/chromeTests.py [hunk 2]
```diff
@@ -58,49 +58,181 @@ def checkbox_labelled_by_inner_element():
...
+			"highlighted",
+			"has details",
+			"cat",  # highlighted content
+			"out of highlighted",
+		]),
+		message="Focus mode: Move by word to word with details"
+	)
+	_asserts.braille_matches(
+		actualBraille,
+		# a backslash character was typed when the gesture was used
```
Comment (seanbudd) [inline tests/system/robot/chromeTests.py:219]: The backslashes seem to be stripped now, I'm guessing this comment is no longer necessary?
Has the behaviour changed, and this no longer happens in NVDA?

File: source/NVDAObjects/IAccessible/__init__.py [hunk 3]
```diff
@@ -1743,6 +1759,37 @@ def __contains__(self, obj):
...
+			ctypes.byref(text),
+			# [in, defaultvalue(FALSE)] const boolean useNewText,
+			False,
+			# [in, defaultvalue(TRUE)] const boolean recurse,
+			True,
+			# [in, defaultvalue(TRUE)] const boolean includeTopLevelText
+			True,
+		)
+		if res != 0:
+			log.error("Error")
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/__init__.py:1789]: Could this be more descriptive?

File: tests/system/robot/chromeTests.py (hunk 2 above)
Comment (feerrenrut) [reply tests/system/robot/chromeTests.py:219]: I missed the comment. If you look at the history of the file, in the earlier commits the backslash was sent to the application because the gesture was only in browse mode.

**Comment by feerrenrut** [comment]
@michaelDCurran I'd like your feedback on the UX, particularly with braille. But also I'd like to hear your thoughts on the [suggestions from @lukaszgo1](https://github.com/nvaccess/nvda/pull/13106#discussion_r759209115)

File: tests/system/robot/chromeTests.py (hunk 2 above)
Comment (seanbudd) [reply tests/system/robot/chromeTests.py:219]: Oh right, that makes sense. There's a few of these still in the document.

**Comment by seanbudd** [comment]
I'd also like an additional reviewer for the C++, I don't feel comfortable reviewing it alone. I'll leave an approval when I am done reviewing everything (and happy), I've had a look at the rest of the PR already.

File: nvdaHelper/interfaces/nvdaInProcUtils/nvdaInProcUtils.idl [hunk 4]
```diff
@@ -74,6 +74,22 @@ interface NvdaInProcUtils {
...
+
+	cpp_quote("/* Actually maps to nvdaInProcUtils_getTextFromIAccessible")
+	cpp_quote("*/")
+	error_status_t getTextFromIAccessible(
+		[in] handle_t bindingHandle,
+		[in] const unsigned long hwnd,
+		[in] long parentID,
+		// Params for getTextFromIAccessible
+		[out, string] BSTR* textBuf,
+		[in, defaultvalue(FALSE)] const boolean useNewText,
```
Comment (michaelDCurran) [inline nvdaHelper/interfaces/nvdaInProcUtils/nvdaInProcUtils.idl:86]: This argument should be removed as it cannot be used cross-process. IAccessibleText::newText is only valid during an in-process winEvent callback. E.g. IA2_EVENT_TEXT_INSERTED. Out-of-process winEvent callbacks are asynchronous and therefore newText will no longer contain the correct information.

File: nvdaHelper/readme.md [hunk 5]
```diff
@@ -71,3 +71,15 @@ When `fillVBuf` is called, it recursively descends through the child elements th
 `controlFieldNode`s or `textFieldNode`s for them.
 This code is responsible for interacting with the IA2 accessibility API, these calls should be minimised for
  performance reasons.
+
+
+### Overview
+
+- NVDA's python code calls into the (`NVDAHelperLocal.dll` / `NVDAHelperLocalWin10.dll`).
+- Generated RPC Wrappers are called.
```
Comment (michaelDCurran) [inline nvdaHelper/readme.md:79]: client-side RPC wrappers are only available in nvdaHelperLocal.dll. The above text seems to suggest that nvdaHelperLocalWin10.dll also contains them which is not correct.

File: source/appModuleHandler.py [hunk 6]
```diff
@@ -345,7 +350,10 @@ def __init__(self,processID,appName=None):
 		#: @type: str
 		self.appName=appName
 		self.processHandle=winKernel.openProcess(winKernel.SYNCHRONIZE|winKernel.PROCESS_QUERY_INFORMATION,False,processID)
-		self.helperLocalBindingHandle=None
+
+		self.helperLocalBindingHandle: Optional[Any] = None
```
Comment (michaelDCurran) [inline source/appModuleHandler.py:354]: I think this is always an Optional[ctypes.c_long]

File: tests/system/libraries/ChromeLib.py [hunk 7]
```diff
@@ -122,6 +122,7 @@ def _waitForStartMarker(self, spy, lastSpeechIndex):
 			actualSpeech = spy.get_speech_at_index_until_now(lastSpeechIndex)
 			if self._wasStartMarkerSpoken(actualSpeech):
 				break
+			spy.emulateKeyPress('F6')
```
Comment (michaelDCurran) [inline tests/system/libraries/ChromeLib.py:125]: why was this added?

**Review by michaelDCurran — CHANGES_REQUESTED** [review]
What about testing the situation where the aria-details is on a span around a link?
```
<p>Hello <span aria-details="xx" role="mark">this is a <a href="https://www.google.com/">test</a></span></p>
```
Will show details summary work when the caret is within the link? I'm not sure it will as NVDAObjectAtStart will just be the link, but details will be on its parent.

File: tests/system/libraries/ChromeLib.py (hunk 7 above)
Comment (feerrenrut) [reply tests/system/libraries/ChromeLib.py:125]: When running locally, and chrome is already open (windowed) the test does not open in kiosk mode. For some reason focus isn't always on the document in that case. I will remove this.

It is a little strange that we have a loop that doesn't change anything between tests now.

File: nvdaHelper/interfaces/nvdaInProcUtils/nvdaInProcUtils.idl (hunk 4 above)
Comment (feerrenrut) [reply nvdaHelper/interfaces/nvdaInProcUtils/nvdaInProcUtils.idl:86]: I'll remove this from the params now and document it. Then in #13127 split this function into multiple responsibilities.

File: nvdaHelper/readme.md (hunk 5 above)
Comment (feerrenrut) [reply nvdaHelper/readme.md:79]: Thanks, I'll try to clarify this further.

File: source/globalCommands.py (hunk 1 above)
Comment (feerrenrut) [reply source/globalCommands.py]: I believe this is now addressed.

File: source/NVDAObjects/IAccessible/__init__.py [hunk 8]
```diff
@@ -1743,6 +1759,35 @@ def __contains__(self, obj):
 		except COMError:
 			return False
 
+	def summarizeInProcess(self) -> str:
+		"""Uses nvdaInProcUtils to get the text for an IAccessible.
+		Can be used without a virtual buffer loaded.
+		"""
+		from comtypes import BSTR
```
Comment (lukaszgo1) [inline source/NVDAObjects/IAccessible/__init__.py:1766]: Any particular reason why these are imported lazily?

File: source/globalCommands.py [hunk 9]
```diff
@@ -1956,6 +1961,71 @@ def script_reportOrShowFormattingAtCaret(self, gesture):
...
+		# controlIdentifier_ID) the docHandle will need to be fetched another way.
+		# Example:
+		#
+		# docHandle = f.field.get("controlIdentifier_docHandle")
+		# ID = f.field.get("controlIdentifier_ID")
+		# objWithDetails = IAccessible.getNVDAObjectFromEvent(docHandle, winUser.OBJID_CLIENT, ID)
+		annotation: Optional[str] = nvdaObject.detailsSummary
+		if annotation:
+			log.debug("NVDAObjectAtStart of caret has details.")
+		elif api.getNavigatorObject():
```
Comment (lukaszgo1) [inline source/globalCommands.py:2012]: In which cases is it necessary to fall back to the navigator objects and why you prefer to use it rather than `api.getFocusObject`?

File: tests/system/robot/chromeTests.py [hunk 10]
```diff
@@ -55,52 +55,306 @@ def checkbox_labelled_by_inner_element():
...
+REVIEW_CURSOR_FOLLOW_CARET_KEY = ["reviewCursor", "followCaret"]
+
+
 def test_mark_aria_details():
+	spy = _NvdaLib.getSpyLib()
+	spy.set_configValue(REVIEW_CURSOR_FOLLOW_CARET_KEY, True)
+	exercise_mark_aria_details()
+
+
+def test_mark_aria_details_FreeReviewCursor():
```
Comment (lukaszgo1) [inline tests/system/robot/chromeTests.py:67]: Thanks for introducing this additional test, however in most cases people who set review cursor not to follow the caret also disable navigator object following focus. Would it be possible to introduce a test making sure that in this case the behavior stays as expected?

File: nvdaHelper/remote/textFromIAccessible.h [hunk 11]
```diff
@@ -16,6 +16,16 @@ This license can be found at:
 #define IA2TEXTFROMIACCESSIBLE_H
 struct IAccessible2;
 
+/* getTextFromIAccessible
+* Summarizes the text for an IAccessible
+* @param textBuf: An out-param, the string to fill with the text from the IAccessible
+* @param pacc2: The IAccessible2 object to get the text from.
+* @param useNewText: Only valid during an in-process winEvent callback (ie text changed). Specifies
+*    to use the new text (ie post change) should be used instead of the old text.
```
Comment (michaelDCurran) [inline nvdaHelper/remote/textFromIAccessible.h:24]:
```suggestion
*    to use the new text (ie only the inserted/changed text) should be used instead of all the current text in the Accessible. 
```

File: source/globalCommands.py (hunk 9 above)
Comment (michaelDCurran) [reply source/globalCommands.py:2012]: I agree. If caret position is used instead of review position, then focus object should be used instead of navigator object.

**Review by michaelDCurran — CHANGES_REQUESTED** [review]

File: source/globalCommands.py (hunk 9 above)
Comment (feerrenrut) [reply source/globalCommands.py:2012]: So that I can add a comment here about why, do either of you have an explanation?

File: tests/system/robot/chromeTests.py (hunk 10 above)
Comment (feerrenrut) [reply tests/system/robot/chromeTests.py:67]: > also disable navigator object following focus

Which setting is that exactly? There is an option for "Review cursor follows focus", but I haven't found any settings about the navigator object following focus.

File: nvdaHelper/remote/textFromIAccessible.h (hunk 11 above)
Comment (feerrenrut) [reply nvdaHelper/remote/textFromIAccessible.h:24]: Is current really the right description? Surely this the event is saying the text has already changed.

File: source/globalCommands.py (hunk 9 above)
Comment (lukaszgo1) [reply source/globalCommands.py:2012]: I'm not sure what you're asking here. If the question  is: "Why should I use focused object rather than navigator object?" then the answer has been given by @michaelDCurran  above. However if you're asking "Why use a fallback when there are no annotations at the caret position?" then I cannot answer that one and was hoping you can explain in which situations there were no annotations at the caret position yet they were present on the navigator object.

File: tests/system/robot/chromeTests.py (hunk 10 above)
Comment (lukaszgo1) [reply tests/system/robot/chromeTests.py:67]: I meant 'Review cursor follows focus'. Sorry for the confusion I didn't check how this option is named in the GUI.

Comment (feerrenrut) [reply tests/system/robot/chromeTests.py:67]: Thanks for clarifying, I thought it might be that, but wanted be sure.

File: source/globalCommands.py (hunk 9 above)
Comment (feerrenrut) [reply source/globalCommands.py:2012]: Do you mean this comment:
>I agree. If caret position is used instead of review position, then focus object should be used instead of navigator object.

This seems to imply that review position and navigator object are some kind of semantic pair, and caret and focus are another semantic pair. Is that true?

I'd like to include the reasoning (the answer to "why is it done this way?") in the code comments. This helps future developers understand the intent, the use-cases, and helps to prevent regressions.

Unfortunately I don't feel like I have a good understanding of this. The comments so far are telling me what this should do without explaining why that is important. 

If I had to guess, it's because of settings like `Review cursor follows focus` which when disabled may make using the navigator object unreliable. Though I don't know that for sure, how is the navigator object related to this setting that talks about the review cursor?

Comment (lukaszgo1) [reply source/globalCommands.py:2012]: > This seems to imply that review position and navigator object are some kind of semantic pair, and caret and focus are another semantic pair. Is that true?

That is one way to look at it. Essentially when moving the review cursor you're moving the navigator object. This is a slight oversimplification, since if you are in object review the navigator object stays the same but in both screen review and document review navigator object follows the review cursor position. The relation between navigator object and review cursor is however not as important as having an understanding as to why we need a fallback at all. Could you please explain why you've written this section of code in the first place. Ideally the code comment here should explain when the fallback (regardless if it is to the focused or navigator object) is needed.

Comment (feerrenrut) [reply source/globalCommands.py:2012]: I agree a better comment is required here, this is supposed to be for the corollary of:
> `# Common cases, vbuf available: Eg editable text, or regular web content`

I.E. This section of code was intended for when the virtual buffer is not available, given the other changes I'm not actually sure it is still required. I believe there is an example when using Google Sheets, however right now I can't recreate it. I'll have to talk to @michaelDCurran after the break about the exact case. We should also add a test for this case.

File: nvdaHelper/remote/textFromIAccessible.h (hunk 11 above)
Comment (michaelDCurran) [reply nvdaHelper/remote/textFromIAccessible.h:24]: I can see why current might be ambiguous, but I am struggling to find a better word. Either way though, "old text" is not correct.
Perhaps I can illustrate with an example and then you can find wording you are happy with.
* Let there be an Accessible object whose text is "abcde", I.e. fetched from IAccessibleText::text.
* Some new text is inserted, and therefore a textInserted event is fired.
* IAccessibleText::text now produces "abxxcde" (I.e. "xx" was inserted after "ab").
* Within the context of the textInserted event, IAccessibleText::newText is "xx". 
So if this function was called within the textInserted event, it would fetch "abxxcde" if useNewText is false, and "xx" if useNewText is true.

Comment (feerrenrut) [reply nvdaHelper/remote/textFromIAccessible.h:24]: Oh I see. Yes, that helps.
It looks like this only applies to the events:
- `IA2_EVENT_TEXT_INSERTED`
- `IA2_EVENT_TEXT_REMOVED`
- `IA2_EVENT_TEXT_UPDATED`.

In the event handlers for those events, `IAccessibleText::text` will reflect the fully realized modification.
`IAccessibleText::newText` is used to determine the additions to the overall text in the case of `IA2_EVENT_TEXT_INSERTED` and `IA2_EVENT_TEXT_UPDATED` and `IAccessibleText::oldText` is used to determine the removals from the overall text in the case of `IA2_EVENT_TEXT_REMOVED` and `IA2_EVENT_TEXT_UPDATED`

When `useNewText` is given, then the fetched text is only the text added, ie we only fetch `IAccessibleText::newText`

**Comment by feerrenrut** [comment]
I have addressed all the feedback and pushed 4 new commits. It turns out we can't find an example where getting the caret position will fail with Chromium / Firefox, theoretically the fallback to rely on the focus object is not necessary, however, I have opted to retain it since there is no negative impact, and the behavior from browsers is somewhat surprising here. I have attempted to add comments to clarify.

File: source/globalCommands.py [hunk 12]
```diff
@@ -1956,6 +1961,57 @@ def script_reportOrShowFormattingAtCaret(self, gesture):
 		elif repeats == 1:
 			self.script_showFormattingAtCaret(gesture)
 
+	@script(
+		description=_(
+			# Translators: the description for the reportDetailsSummary script.
+			"Report summary of any annotation details at the current location."
+		)
+	)
```
Comment (lukaszgo1) [inline source/globalCommands.py:1969]: Without assigning a category to this script it would show in the Miscellaneous category in the input gestures dialog. I'm not against that - just wanted to make sure this was though through rather than done on accident.

File: source/globalCommands.py [hunk 13]
```diff
@@ -1956,6 +1961,57 @@ def script_reportOrShowFormattingAtCaret(self, gesture):
...
 			self.script_showFormattingAtCaret(gesture)
 
+	@script(
+		description=_(
+			# Translators: the description for the reportDetailsSummary script.
+			"Report summary of any annotation details at the current location."
+		)
+	)
+	def script_reportDetailsSummary(self, gesture):
+		"""Report the annotation details summary for the single character under the caret or the current nav object.
```
Comment (lukaszgo1) [inline source/globalCommands.py:1971]: This comment is not valid anymore since we're using the focus object rather than nav object.

File: source/globalCommands.py [hunk 14]
```diff
@@ -1956,6 +1961,57 @@ def script_reportOrShowFormattingAtCaret(self, gesture):
...
+		@note: It is tempting to try to report any annotation details that exists in the range formed by prior
+			and current location. This would be a new paradigm in NVDA, and may feel natural when moving by line
+			to be able to more quickly have the 'details' reported. However, there may be more than one 'details
+			relation' in that range and, we don't yet have a way for the user to select which one to report.
```
Comment (lukaszgo1) [inline source/globalCommands.py:1975]: I'm not a native English speaker but the comma after 'and' looks strange to me.

File: source/globalCommands.py [hunk 15]
```diff
@@ -1956,6 +1961,57 @@ def script_reportOrShowFormattingAtCaret(self, gesture):
...
+			For now, we minimise this risk by only reporting details at the current location.
+		"""
+		log.info("Report annotation details summary at current location.")
```
Comment (lukaszgo1) [inline source/globalCommands.py:1978]: Can this be logged at a lover level or removed completely - I don't think  it adds any value?

File: source/globalCommands.py [hunk 16]
```diff
@@ -1956,6 +1961,57 @@ def script_reportOrShowFormattingAtCaret(self, gesture):
...
+		try:
+			# Common cases use Caret Position: vbuf available or object supports text range
+			# Eg editable text, or regular web content
+			# Firefox and Chromium support this even in a button within a role=application.
+			caret: textInfos.TextInfo = api.getCaretPosition()
+		except ValueError:
```
Comment (lukaszgo1) [inline source/globalCommands.py:1984]: Since `api.getCaretPosition` can also raise `RuntimeError` shouldn't it also be handled here?

File: source/globalCommands.py [hunk 17]
```diff
@@ -1956,6 +1961,57 @@ def script_reportOrShowFormattingAtCaret(self, gesture):
...
+			log.debugWarning("Unable to get the caret position.", exc_info=True)
+			return
+		caret.expand(textInfos.UNIT_CHARACTER)
+		nvdaObject: Optional[NVDAObject] = caret.NVDAObjectAtStart
```
Comment (lukaszgo1) [inline source/globalCommands.py:1988]: I'm not sure in which cases `nvdaObject` can be `None` if we can create a caret position. At any rate if it indeed can be `None` we should not  try to get annotations from it.

File: source/globalCommands.py [hunk 18]
```diff
@@ -1956,6 +1961,57 @@ def script_reportOrShowFormattingAtCaret(self, gesture):
...
+		log.debug(f"Trying with nvdaObject : {nvdaObject}")
+
+		annotation: Optional[str] = nvdaObject.detailsSummary
+		if annotation:
+			log.debug("NVDAObjectAtStart of caret has details.")
+		elif api.getFocusObject() and False:
```
Comment (lukaszgo1) [inline source/globalCommands.py:1994]: A left over from debugging.

File: source/globalCommands.py (hunk 12 above)
Comment (feerrenrut) [reply source/globalCommands.py:1969]: Thanks for the reminder, I hadn't thought too hard about which category it should belong to yet. I think it makes most sense in the system caret category, what do you think?

File: source/globalCommands.py (hunk 13 above)
Comment (feerrenrut) [reply source/globalCommands.py:1971]: Ah, yep. I missed that!

File: source/globalCommands.py (hunk 14 above)
Comment (feerrenrut) [reply source/globalCommands.py:1975]: Yes, good catch.

File: source/globalCommands.py (hunk 16 above)
Comment (feerrenrut) [reply source/globalCommands.py:1984]: Yes, this should have been `RuntimeError`.

File: source/globalCommands.py (hunk 17 above)
Comment (feerrenrut) [reply source/globalCommands.py:1988]: I believe this was an annotation left over from a prior approach.

File: source/globalCommands.py (hunk 15 above)
Comment (feerrenrut) [reply source/globalCommands.py:1978]: I'll make it debug level. The main value is easily being able to see when the script starts.

File: source/globalCommands.py (hunk 18 above)
Comment (feerrenrut) [reply source/globalCommands.py:1994]: Good catch.

File: source/globalCommands.py (hunk 12 above)
Comment (lukaszgo1) [reply source/globalCommands.py:1970]: > I hadn't thought too hard about which category it should belong to yet. I think it makes most sense in the system caret category, what do you think?

Agreed.

**Review by michaelDCurran — APPROVED** [review]
